crates/metal-llvm/src/stdlib_priv.rs

```diff
@@ -0,0 +1,96 @@
+use std::{ffi::CString, ptr};
+
+use llvm_sys::{
+    core::LLVMCreateMemoryBufferWithMemoryRangeCopy, ir_reader::LLVMParseIRInContext,
+    linker::LLVMLinkModules2, prelude::*,
+};
+
+// NOTE: private modules do NOT have submodules.
+/// private modules stand at the core of the stdlib.
+/// They allow us to define metal functionality in the stdlib
+/// for compile time.
+pub trait PrivateSTDLibModule {
+    // private modules usually start with an underscore
+    // But this name does not include that underscore.
+    /// Private STDLib module name.
+    /// i.e. `io`, `cmplib`, etc.
+    fn libname() -> &'static str;
+
+    /// # Safety
+    /// The LLVM Pointer values here can cause crashes
+    /// if set null or otherwise. Basically, only use
+    /// if you know what you're doing.
+    unsafe fn function(
+        ident: metal_ast::Ident,
+        ctx: LLVMContextRef,
+        module: LLVMModuleRef,
+        arguments: Vec<String>,
+    ) -> Option<LLVMValueRef>;
+}
+
+pub struct ExternalLib;
+
+impl PrivateSTDLibModule for ExternalLib {
+    fn libname() -> &'static str {
+        "pext"
+    }
+
+    unsafe fn function(
+        ident: metal_ast::Ident,
+        ctx: LLVMContextRef,
+        module: LLVMModuleRef,
+        arguments: Vec<String>,
+    ) -> Option<LLVMValueRef> {
+        match ident.inner {
+            "extern" => {
+                if arguments.len() != 3 {
+                    panic!("extern accepts exactly 3 arguments")
+                }
+                let name = &arguments[0];
+                let ret_type = &arguments[1];
+                let args = &arguments[2];
+
+                let irstr = format!(
+                    "
+                    define {} @{}({}) unnamed_addr
+                ",
+                    ret_type, name, args
+                );
+                let ir = CString::new(irstr.clone()).unwrap();
+
+                unsafe {
+                    let irbuf = CString::new("ir_buffer").unwrap();
+                    let ir_buffer = LLVMCreateMemoryBufferWithMemoryRangeCopy(
+                        ir.as_ptr(),
+                        irstr.len(),
+                        irbuf.as_ptr(),
+                    );
```

Comment:
> `CString::new(irstr.clone())`
> `irstr.len(),`

I believe you should instead pass `ir.len()` instead, as the first and the second arguments represent one "object". That way the clone will also become unnecessary.